### PR TITLE
feat(credentials): store and processing generic app credentials

### PR DESCRIPTION
**Pull request by s1fr0**
This PR adds a dynamic keystore for membership credential storage and management. The credentials are thought to be generic membership credentials, i.e. associated to a membership contract and membership Merkle tree, and the implementation is not tied to RLN credentials only.

The keystore JSON schema implemented is the one proposed in https://github.com/waku-org/nwaku/issues/1238#issuecomment-1278114079. 

In short, a keystore is uniquely identified by an application name, app identifier and version strings. Multiple keystores can co-exist in the same file.
Each keystore can contain none or more Membership credentials. Each Membership Credential contains a unique Identity Credential (made of identity trapdoor, nullifier, secret hash and commitment) and none or more Membership Groups. Each Membership Group contains a Membership contract and a membership tree index position. A Membership Contract contains the chain ID and the membership contract address information.

When a Membership credential is added to a certain keystore: 
- if its Identity Credential was present in the keystore (it decrypts with the same password used to add the new credential), only the new Membership Groups (i.e. not already present in the keystore) will be added;
- if not present, the whole credential is added.
When one or more credentials are added, the keystore is automatically updated to disk (with a backup mechanism to mitigate credentials loss in case of IO errors).

Membership credentials in a certain keystore can be retrieved by none or more filters. These allow to return:
- all Membership credentials decrypting under a certain password;
- Membership credentials corresponding to one or more Identity credentials;
- Membership credentials corresponding to one or more Membership Groups (non-matching groups are stripped);
- Membership credentials corresponding to one or more Identity credentials and Membership Groups (non-matching groups are stripped).

This PR addresses the issues:
  - [x] https://github.com/vacp2p/rfc/issues/543
  - [x] https://github.com/status-im/nwaku/issues/1238
  - [x] https://github.com/status-im/nwaku/issues/1278

### CODE REVIEW

**Comment by jm-clius** [comment]
@s1fr0 @rymnc I notice this draft PR has a release milestone attached - should we remove the milestone (not required for PRs, but rather for issues), or bump it to a next milestone?

File: waku/v2/node/waku_node.nim [hunk 1]
```diff
@@ -285,7 +285,7 @@ proc info*(node: WakuNode): WakuInfo =
 proc connectToNodes*(node: WakuNode, nodes: seq[RemotePeerInfo] | seq[string], source = "api") {.async.} =
   ## `source` indicates source of node addrs (static config, api call, discovery, etc)
   # NOTE This is dialing on WakuRelay protocol specifically
-  await connectToNodes(node.peerManager, nodes, WakuRelayCodec, source=source)
+  await peer_manager.connectToNodes(node.peerManager, nodes, WakuRelayCodec, source=source)
```
Comment (rymnc) [inline waku/v2/node/waku_node.nim:288]: Is this change required?

File: waku/v2/utils/credentials.nim [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+  # hash of user's identity key generated by
+  # Poseidon hash function implemented in rln lib
+  # more details in https://hackmd.io/tMTLMYmTR5eynw2lwK9n1w?view#Membership
+  idCommitment*: IDCommitment
+
+type MembershipIndex* = uint
+
+type MembershipContract* = object
+  chainId*: string
+  address*: string
```
Comment (rymnc) [inline waku/v2/utils/credentials.nim:35]:
```suggestion
type ContractReference* = object
  chainId*: string
  address*: string
```

File: waku/v2/utils/credentials.nim [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+    err(OsError)
+  finally:
+    f.close()
+
+# This proc load a keystore based on the application, appIdentifier and version filters. 
+# If none is found, it automatically creates an empty keystore for the passed parameters
+proc loadAppKeystore*(path: string,
+                      application: string,
+                      appIdentifier: string,
+                      version: string,
```
Comment (rymnc) [inline waku/v2/utils/credentials.nim:124]: Can this be a single object? `AppConfig` or similar

File: waku/v2/utils/credentials.nim [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+      try:
+        # We parse the json
+        data = json.parseJson(keystore)
+
+        # We check if parsed json contains the relevant keystore credentials fields and if these are set to the passed parameters 
+        # (note that "if" is lazy, so if one of the .contains() fails, the json fields contents will not be checked and no ResultDefect will be raised due to accessing unavailable fields)
+        if data.contains("application") and data.contains("appIdentifier") and data.contains("credentials") and data.contains("version") and
+           data["application"].getStr() == application and 
+           data["appIdentifier"].getStr() == appIdentifier and 
+           data["version"].getStr() == version:
```
Comment (rymnc) [inline waku/v2/utils/credentials.nim:164]: Can you create a proc which checks if `data` contains all of the elements in ["application", "appIdentifier", "credentials", "version"]?

```nim
proc hasAllKeys(data: Table[string, JsonNode], keys: seq[string]): bool =
  return all(data.contains(key) : key in keys)
```

File: waku/v2/utils/credentials.nim [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+      # A flag to tell us if the keystore contains a credential associated to the input identity credential, i.e. membershipCredential
+      found = false
+
+      for keystoreCredential in keystoreCredentials.mitems():
+        # keystoreCredential is encrypted. We decrypt it
+        let decodedKeyfile = decodeKeyFileJson(keystoreCredential, password)
+        if decodedKeyfile.isOk():
+
+          # we parse the json decrypted keystoreCredential
+          let decodedCredential = decode(decodedKeyfile.get())
```
Comment (rymnc) [inline waku/v2/utils/credentials.nim:223]: For all calls which return a `Result`, please append `Res` to the variable

File: waku/v2/utils/credentials.nim [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+        return err(OsError)
+
+  except IOError:
+    return err(IoError)
+
+  return ok(matchingAppKeystore)
+
+
+# Adds a sequence of membership credential to the keystore matching the application, appIdentifier and version filters. 
+proc addMembershipCredentials*(path: string,
```
Comment (rymnc) [inline waku/v2/utils/credentials.nim:188]: Can we shorten this proc? It's > 100 loc :)

**Review by rymnc — COMMENTED** [review]
Few queries, looks great!

Maybe we should decide if we plan to merge this, or #1496 first. There would be non-trivial refactoring both ways.

**Comment by LNSD** [comment]
Wow! +696 new lines! 😨

I will need a couple of days to review this. Let's keep the PR sizes manageable. Smaller PRs will help us be more agile in reviewing and merging code changes.

File: waku/v2/utils/credentials.nim (hunk 2 above)
Comment (rymnc) [reply waku/v2/utils/credentials.nim:35]: No strong opinion here, but MembershipContract can refer to the type of the contract itself (generated by nim-web3). This works too!

**Review by rymnc — APPROVED** [review]
LGTM, maybe we should have instructions for operators to use this feature :)

File: waku/v2/utils/credentials.nim [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+    OsError               = "keystore error: OS specific error"
+    IoError               = "keystore error: IO specific error"
+    JsonKeyError          = "keystore error: fields not present in JSON"
+    JsonError             = "keystore error: JSON encoder/decoder error"
+    KeystoreDoesNotExist  = "keystore error: file does not exist"
+    CreateKeystoreError   = "Error while creating application keystore"
+    LoadKeystoreError     = "Error while loading application keystore"
+    CreateKeyfileError    = "Error while creating keyfile for credentials"
+    SaveKeyfileError      = "Error while saving keyfile for credentials"
+    ReadKeyfileError      = "Error while reading keyfile for credentials"
```
Comment (LNSD) [inline waku/v2/utils/credentials.nim:67]: Could you separate the `AppKeystore` related code into another module?

File: waku/v2/utils/credentials.nim [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+    let jsonObject = parseJson(string.fromBytes(encodedCredential))
+    return ok(to(jsonObject, MembershipCredentials))
+  except JsonParsingError:
+    return err(JsonError)
+  except Exception: #parseJson raises Exception
+    return err(OsError)
+
+# Checks if a JsonNode has all keys contained in "keys"
+proc hasKeys*(data: JsonNode, keys: openArray[string]): bool =
+  return all(keys, proc (key: string): bool = return data.hasKey(key))
```
Comment (LNSD) [inline waku/v2/utils/credentials.nim:93]: All the JSON serialization-related code should be in a separate module.

**Review by LNSD — CHANGES_REQUESTED** [review]
I find the code a little convoluted. Please, check my comments.

**Review by LNSD — APPROVED** [review]
Thanks for adding it inside the `waku_keystore`. I like the new module layout.

There are minor/nitpick details, but in general, LGreatTM ✅

**Pull request merged**